### PR TITLE
loookup-table impl (covers proposal)

### PR DESCRIPTION
**Pull request by janmejay**
lookup-table proposal (http://www.rsyslog.com/doc/lookup_tables.html) implemented

### CODE REVIEW

**Comment by janmejay** [comment]
This doesn't yet include load_lookup_table (should have mentioned it in original description, but better late than never). I somehow missed it altogether in the list of things to implement. 

Will implement is on the same PR if it remains open for a few more days, if merged in the meanwhile, will open a new PR on this.

There is no need to keep this PR open just for the load_lookup_table change though (because its small and independent change, which can be merged separately).

**Comment by rgerhards** [comment]
No problem at all. As this needs quite some review, I expect that the merge takes some while :-)

**Comment by janmejay** [comment]
@rgerhards how about implementing load_lookup_table as a function? The reset-failure can in that case be handled in the best way (because user can then choose what to do when it fails).

Infact, if we choose to, we can implement another statement "stub_lookup_value(value)" which will substitute lookup_table with a dummy table, which always returns given value. reload_lookup_table function failure can be checked by user and when desired, stub_lookup_table function can be called for failure.

Eg.

    if (not reload_lookup_table("foo")) then {
      stub_lookup_value("reload-failed");
    }

Thoughts?

**Comment by janmejay** [comment]
I propose this because processBatch doesn't do error-checking on statements, because of which even though reload_lookup_table will return error-code, user won't have a way of reacting to it. 

Of-course we can bake-in some reactions in the statement/action (eg. valueOnFail param), but its still doesn't have the same level of flexibility.

**Comment by janmejay** [comment]
I can fix processBatch to propagate errors and abort further processing, but that will affect existing users that expect failing statements not to stop processing.

**Comment by rgerhards** [comment]
Stricly speaking, we do not have any existing users. And even those that have found this non-existing feature are few. And I guess the read release notes very carefully. So I think we can make this change as long as we properly document it and also send a note out on the rsyslog mailing list.

Note that I will not merge for the 8.15 release frame, as I am working on some more stabilizations of the json locking issues we found inside the core json processing. I don't want to have any other larger changes for that release ... and I also admit I don't have time for the in-depth review required in december. So you have ample time to improve what you thinks makes sense to improve.

**Comment by janmejay** [comment]
Cool, I'll socialize implementing it as reload_lookup_table(table_name)
function and stub_lookup_table_value(value) statement.

Will send a mail in some time.


On Thu, Nov 26, 2015 at 5:09 PM, Rainer Gerhards <notifications@github.com>
wrote:

> Stricly speaking, we do not have any existing users. And even those that
> have found this non-existing feature are few. And I guess the read release
> notes very carefully. So I think we can make this change as long as we
> properly document it and also send a note out on the rsyslog mailing list.
>
> Note that I will not merge for the 8.15 release frame, as I am working on
> some more stabilizations of the json locking issues we found inside the
> core json processing. I don't want to have any other larger changes for
> that release ... and I also admit I don't have time for the in-depth review
> required in december. So you have ample time to improve what you thinks
> makes sense to improve.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-159891302>.
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by janmejay** [comment]
WRT merge-timeline, no problem at all.

On Thu, Nov 26, 2015 at 5:38 PM, singh.janmejay <singh.janmejay@gmail.com>
wrote:

> Cool, I'll socialize implementing it as reload_lookup_table(table_name)
> function and stub_lookup_table_value(value) statement.
>
> Will send a mail in some time.
>
>
> On Thu, Nov 26, 2015 at 5:09 PM, Rainer Gerhards <notifications@github.com
> > wrote:
>
>> Stricly speaking, we do not have any existing users. And even those that
>> have found this non-existing feature are few. And I guess the read release
>> notes very carefully. So I think we can make this change as long as we
>> properly document it and also send a note out on the rsyslog mailing list.
>>
>> Note that I will not merge for the 8.15 release frame, as I am working on
>> some more stabilizations of the json locking issues we found inside the
>> core json processing. I don't want to have any other larger changes for
>> that release ... and I also admit I don't have time for the in-depth review
>> required in december. So you have ample time to improve what you thinks
>> makes sense to improve.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-159891302>.
>>
>
>
>
> --
> Regards,
> Janmejay
> http://codehunk.wordpress.com
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by portante** [comment]
@janmejay, we have a usecase in mind for this, where I want to normalize a set of fields, for example, mapping "SEVERE" to "err", "AUDIT" to "debug", etc. for log levels, or other anomalies in the various fields.  These often occur on legacy systems where we can't fix them up easily, and we'd like to be able to handle those values at the point of aggregation.

If you end up with a working patch, please include me when making it available.

**Comment by janmejay** [comment]
All of that is already implemented in PR. Its also been available in
Rsyslog releases made 2013 (second half) onwards (it wasn't a complete
implementation, this completes the proposal).

You likely don't need to apply any patches, it'll just work.

On Thu, Nov 26, 2015 at 9:18 PM, Peter Portante <notifications@github.com>
wrote:

> @janmejay <https://github.com/janmejay>, we have a usecase in mind for
> this, where I want to normalize a set of fields, for example, mapping
> "SEVERE" to "err", "AUDIT" to "debug", etc. for log levels, or other
> anomalies in the various fields. These often occur on legacy systems where
> we can't fix them up easily, and we'd like to be able to handle those
> values at the point of aggregation.
>
> If you end up with a working patch, please include me when making it
> available.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-159945411>.
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by janmejay** [comment]
The rscript based table-reload and stub-value impl done. Reload is implemented as a function and stub-value as a statement. 

I had to cherrypick some of the patches from dyn-stats PR to get valgrind on travis-machines to accept string operations.

The PR is ready for merge now.

**Comment by janmejay** [comment]
Implemented async rscript based table-reload with optional stub-value which will be used in case reload fails.

Killed the separate stub statement because the good-composition way of doing it involved extracting result from reload-function, which required waiting for reload to complete, which wasn't a workable option for very large tables. 

So PR now follows semantics described in the original proposal.

Finally ready for merge.

**Comment by rgerhards** [comment]
@janmejay in dynstats.c, for function

```static void
dynstats_readCallback(statsobj_t *ignore, void *b) {```

parameter "ignore" is not used. Can we remove it? As a work-around, I'll first assign unused attribute.

**Comment by janmejay** [comment]
It is to match callback interface. statsobj subsystem allows statsobj_t
owner to get an optional callback once stats are read. That parameter
allows a callback function to recognize which statsobj has been read, in
case it is dealing with multiple.

Its a convenience, technically user can keep statsobj_t reference in
context object anyway. But in some scenarios, it'll simplify code (callback
won't have to fold that reference into context).

I think you wanted to post this on dyn-stats PR though.

On Thu, Jan 28, 2016 at 2:32 PM, Rainer Gerhards <notifications@github.com>
wrote:

> @janmejay <https://github.com/janmejay> in dynstats.c, for function
>
> static void
> dynstats_readCallback(statsobj_t *ignore, void *b) {
>
> parameter "ignore" is not used. Can we remove it? As a work-around, I'll
> first assign unused attribute.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-176071941>.
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by janmejay** [comment]
But if you feel we should kill it, it'll require killing the param here and
then changing callback signature in statsobj.

On Thu, Jan 28, 2016 at 2:50 PM, singh.janmejay <singh.janmejay@gmail.com>
wrote:

> It is to match callback interface. statsobj subsystem allows statsobj_t
> owner to get an optional callback once stats are read. That parameter
> allows a callback function to recognize which statsobj has been read, in
> case it is dealing with multiple.
>
> Its a convenience, technically user can keep statsobj_t reference in
> context object anyway. But in some scenarios, it'll simplify code (callback
> won't have to fold that reference into context).
>
> I think you wanted to post this on dyn-stats PR though.
>
> On Thu, Jan 28, 2016 at 2:32 PM, Rainer Gerhards <notifications@github.com
> > wrote:
>
>> @janmejay <https://github.com/janmejay> in dynstats.c, for function
>>
>> static void
>> dynstats_readCallback(statsobj_t *ignore, void *b) {
>>
>> parameter "ignore" is not used. Can we remove it? As a work-around, I'll
>> first assign unused attribute.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-176071941>.
>>
>
>
>
> --
> Regards,
> Janmejay
> http://codehunk.wordpress.com
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by rgerhards** [comment]
opps... indeed, selected the wrong PR :-(

If it's part of the interface spec, I think "attribute unused" is fine here. Now on to our regular programming ;)

**Comment by rgerhards** [comment]
@janmejay In order to speed up merging, I think about flagging the feature as "experimental/not yet fully supported for production". This would relieve me from going through all the code before merging. any concerns?

**Comment by janmejay** [comment]
Sounds good to me.

--
Regards,
Janmejay

PS: Please blame the typos in this mail on my phone's uncivilized soft
keyboard sporting it's not-so-smart-assist technology.

**Comment by rgerhards** [comment]
@janmejay would you mind doing a rebase to current master? It's not stricly necessary but would definitely ease my life and merge process ;)

**Comment by rgerhards** [comment]
as a side-note, I think we want to merge this into master now, right?

**Comment by janmejay** [comment]
Sure, I'll rebase this to master today.

Oh, did you mean between master and master-insecure?  I thought you'll do 2
merges (first: my branch -> master-insecure, then master-insecure ->
master). That is why I raised PR towards master-insecure.

I was expecting merge to master to be complex at that time, because I was
expecting lookup_table code to be gone and incremental changes to be
difficult to put in place in absence of the original code.

But because deletion didn't happen, it should be similar in complexity.

Eventually it should end up in master, either way is fine.

On Wed, Feb 3, 2016 at 11:38 PM, Rainer Gerhards <notifications@github.com>
wrote:

> as a side-note, I think we want to merge this into master now, right?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-179382077>.
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by janmejay** [comment]
Quick question: will merge do?

I'm thinking of merging master into this branch. That way incremental
conflicts will not need to be resolved.

--
Regards,
Janmejay

PS: Please blame the typos in this mail on my phone's uncivilized soft
keyboard sporting it's not-so-smart-assist technology.

**Comment by rgerhards** [comment]
Should have thought of this. Yes, merge will do perfectly.

**Comment by rgerhards** [comment]
on master vs. master-insecure: could be that my orginal plans changed. But as you say, it's not really a difference. I just fast-forward master-insecure to master, so it's actually equal right at this moment.

The original intent, as you say, was to have master-insecure in place if we were forced to temporarily remove lookup table functionality. Otherwise, they should always have stayed in sync (but I didn't bother to update). Hope that clarifies.

In other words: just leave the PR to master-insecure, I'll do the second merge once it is in (actually three, as I need to go through master-candidate for testing purposes ;)).

**Comment by janmejay** [comment]
Pushed the pre-merge and fixed issues that static-analysis pointed out.


On Thu, Feb 4, 2016 at 12:39 PM, Rainer Gerhards <notifications@github.com>
wrote:

> on master vs. master-insecure: could be that my orginal plans changed. But
> as you say, it's not really a difference. I just fast-forward
> master-insecure to master, so it's actually equal right at this moment.
>
> The original intent, as you say, was to have master-insecure in place if
> we were forced to temporarily remove lookup table functionality. Otherwise,
> they should always have stayed in sync (but I didn't bother to update).
> Hope that clarifies.
>
> In other words: just leave the PR to master-insecure, I'll do the second
> merge once it is in (actually three, as I need to go through
> master-candidate for testing purposes ;)).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-179682444>.
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by janmejay** [comment]
Wait, I think GCC build has some problem. I'll look at it.

On Thu, Feb 4, 2016 at 6:59 PM, singh.janmejay <singh.janmejay@gmail.com>
wrote:

> Pushed the pre-merge and fixed issues that static-analysis pointed out.
>
>
> On Thu, Feb 4, 2016 at 12:39 PM, Rainer Gerhards <notifications@github.com
> > wrote:
>
>> on master vs. master-insecure: could be that my orginal plans changed.
>> But as you say, it's not really a difference. I just fast-forward
>> master-insecure to master, so it's actually equal right at this moment.
>>
>> The original intent, as you say, was to have master-insecure in place if
>> we were forced to temporarily remove lookup table functionality. Otherwise,
>> they should always have stayed in sync (but I didn't bother to update).
>> Hope that clarifies.
>>
>> In other words: just leave the PR to master-insecure, I'll do the second
>> merge once it is in (actually three, as I need to go through
>> master-candidate for testing purposes ;)).
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-179682444>.
>>
>
>
>
> --
> Regards,
> Janmejay
> http://codehunk.wordpress.com
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by janmejay** [comment]
GCC build now passes.

**Comment by janmejay** [comment]
@rgerhards this is ready for merge.

**Comment by rgerhards** [comment]
thanks a ton, so let's give it a final test on the multi-platform testbench...

**Comment by rgerhards** [comment]
side-note: solaris currently always fails because it does not yet have libfastjson

**Comment by janmejay** [comment]
Please let me know if it surfaces any problems.

On Fri, Feb 5, 2016 at 2:57 PM, Rainer Gerhards <notifications@github.com>
wrote:

> side-note: solaris currently always fails because it does not yet have
> libfastjson
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-180265464>.
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by rgerhards** [comment]
Unfortunately, there are issues, but not with your code. Something in the buildbot environment is screwed I guess. I have been away most of the day and will try to work something out over the weekend. As far as I can see, it is "merge OK" if once we get past these quirks.

**Comment by janmejay** [comment]
Ok, please do let me know if I can help.

On Fri, Feb 5, 2016 at 10:59 PM, Rainer Gerhards <notifications@github.com>
wrote:

> Unfortunately, there are issues, but not with your code. Something in the
> buildbot environment is screwed I guess. I have been away most of the day
> and will try to work something out over the weekend. As far as I can see,
> it is "merge OK" if once we get past these quirks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rsyslog/rsyslog/pull/578#issuecomment-180455479>.
>



-- 
Regards,
Janmejay
http://codehunk.wordpress.com

**Comment by rgerhards** [comment]
OK, I went a bit further. Now I guess we have a missing API on freebsd: http://build.rsyslog.com:8010/builders/rsyslog%20freebsd%20rsyslog/builds/250/steps/shell_2/logs/stdio

**Comment by janmejay** [comment]
I'll take care of it.

--
Regards,
Janmejay

PS: Please blame the typos in this mail on my phone's uncivilized soft
keyboard sporting it's not-so-smart-assist technology.

**Comment by janmejay** [comment]
@rgerhards The commit time is off because of my local clock skew. Fn pthread_setname_np dependence is now optional (it falls-back to not setting the name in absence of it).

**Comment by rgerhards** [comment]
finally looks great! Thanks again.
